packages/@aws-cdk/aws-elasticsearch/lib/domain.ts

```diff
@@ -0,0 +1,1503 @@
+    }
+
+    if (isInstanceType('t2.micro') && elasticsearchVersionNum > 2.3) {
+      throw new Error('The t2.micro.elasticsearch instance type supports only Elasticsearch 1.5 and 2.3.');
+    }
+
+    // Only R3 and I3 support instance storage, per
+    // https://aws.amazon.com/elasticsearch-service/pricing/
+    if (!ebsEnabled && !isEveryInstanceType('r3', 'i3')) {
+      throw new Error('EBS volumes are required for all instance types except R3 and I3.');
+    }
+
+    // Fine-grained access control requires node-to-node encryption, encryption at rest,
+    // and enforced HTTPS.
+    if (advancedSecurityEnabled) {
+      if (!nodeToNodeEncryptionEnabled) {
+        throw new Error('Node-to-node encryption is required when fine-grained access control is enabled.');
+      }
+      if (!encryptionAtRestEnabled) {
+        throw new Error('Encryption-at-rest is required when fine-grained access control is enabled.');
+      }
+      if (!enforceHttps) {
+        throw new Error('Enforce HTTPS is required when fine-grained access control is enabled.');
+      }
+    }
+
+    let cfnVpcOptions: CfnDomain.VPCOptionsProperty | undefined;
+    if (props.vpcOptions) {
+      cfnVpcOptions = {
+        securityGroupIds: props.vpcOptions.securityGroups.map((sg) => sg.securityGroupId),
+        subnetIds: props.vpcOptions.subnets.map((subnet) => subnet.subnetId),
+      };
+    }
+
+    // Setup logging
+    const logGroups: logs.ILogGroup[] = [];
+
+    if (props.logging?.slowSearchLogEnabled) {
+      this.slowSearchLogGroup = props.logging.slowSearchLogGroup ??
+        new logs.LogGroup(scope, 'SlowSearchLogs', {
+          retention: logs.RetentionDays.ONE_MONTH,
+        });
+
+      logGroups.push(this.slowSearchLogGroup);
+    };
+
+    if (props.logging?.slowIndexLogEnabled) {
+      this.slowIndexLogGroup = props.logging.slowIndexLogGroup ??
+        new logs.LogGroup(scope, 'SlowIndexLogs', {
+          retention: logs.RetentionDays.ONE_MONTH,
+        });
+
+      logGroups.push(this.slowIndexLogGroup);
+    };
+
+    if (props.logging?.appLogEnabled) {
+      this.appLogGroup = props.logging.appLogGroup ??
+        new logs.LogGroup(scope, 'AppLogs', {
+          retention: logs.RetentionDays.ONE_MONTH,
+        });
+
+      logGroups.push(this.appLogGroup);
+    };
+
+    let logGroupResourcePolicy: LogGroupResourcePolicy | null = null;
+    if (logGroups.length > 0) {
+      const logPolicyStatement = new iam.PolicyStatement({
+        effect: iam.Effect.ALLOW,
+        actions: ['logs:PutLogEvents', 'logs:CreateLogStream'],
+        resources: logGroups.map((lg) => lg.logGroupArn),
+        principals: [new iam.ServicePrincipal('es.amazonaws.com')],
+      });
+
+      // Use a custom resource to set the log group resource policy since it is not supported by CDK and cfn.
+      // https://github.com/aws/aws-cdk/issues/5343
+      logGroupResourcePolicy = new LogGroupResourcePolicy(this, 'ESLogGroupPolicy', {
```

Comment:
Just for my own knowledge, how does this actually work? How come the policy doesn't need to be passed anywhere? Just created?

Reply:
It's a wrapper for a custom resource construct. So when the stack is run it will make an API call to create a CloudWatch resource policy that allows `es.amazonaws.com` access to the log groups for this cluster.

We also add this as a dependency to the domain to ensure it is run first otherwise the domain creation fails.

Reply:
So basically CloudWatch is treated as a resource that we can add policies to, just like a bucket.

Ok, thanks.